oneflow/user/ops/reshape_user_op_util.cpp

```diff
@@ -106,28 +106,42 @@ Maybe<void> ReshapeUserOpUtil::GetGroupStartInAxis2OutAxis(
       << Error::RuntimeError()
       << "The element number of input tensor must be equal to output tensor, "
       << "but got " << in_shape.elem_cnt() << " and " << out_shape.elem_cnt();
-  int in_axis = in_shape.NumAxes() - 1;
-  int out_axis = out_shape.NumAxes() - 1;
-  while (in_axis >= 0 && out_axis >= 0) {
-    if (in_shape.Count(in_axis) < out_shape.Count(out_axis)) {
-      --in_axis;
-    } else if (in_shape.Count(in_axis) > out_shape.Count(out_axis)) {
-      --out_axis;
-    } else {
+  // Initialization
+  // shape_count is the product of the axis length in [start_axis, end)
+  int64_t in_shape_count = 1;
+  int64_t out_shape_count = 1;
+  int64_t in_axis = in_shape.NumAxes();
+  int64_t out_axis = out_shape.NumAxes();
+  // Move forward functions
+  auto Move2NextAxis = [](const Shape& shape, int64_t& axis, int64_t& shape_count) {
```

Comment:
这里的引用换成指针吧

Reply:
> > The current algorithm has a bug under such situation: In shape: (49, 49, 24) Out shape: (2401, 24) The old algorithm would allow {in: S0, out: S0} on 2 GPUs since 24 is divisible by 2. However, on GPU 0, In physical shape: (25, 49, 24) Out physical shape: (1201, 24) It would cause a bug: `Check failed: (out_shape->elem_cnt()) == (in_shape.elem_cnt()) (28824 vs 29400)` Similar case on GPU 1.
> > This algorithm fixes this bug.
> > With these reasonable SBP candidates, the throughput of auto parallel would increase by 50%.
> 
> 这个 case 需要加到 单测里。

正好把https://github.com/Oneflow-Inc/OneTeam/issues/1637 的代码偷过来